File: Stacks/739. Daily Temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        stack = []
        answer = [0] * n

        for i in range(n):
            while stack and temperatures[stack[-1]] < temperatures[i]:
                j = stack.pop()
                answer[j] = i - j
            stack.append(i)

        return answer

# Time: O(n)
# Space: O(n)

File: Stacks/739. Daily Temperatures/739. Daily Temperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function (temperatures) {
  let stack = [];
  const ans = new Array(temperatures.length).fill(0);

  for (let i = 0; i < temperatures.length; i++) {
    while (stack && temperatures[stack[stack.length - 1]] < temperatures[i]) {
      j = stack.pop();
      ans[j] = i - j;
    }
    stack.push(i);
  }

  return ans;
};

// Time: O(n)
// Space: O(n)
